include/math/Tensor.hpp

```diff
@@ -0,0 +1,245 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// 
+// GQCG-gqcp is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with GQCG-gqcp.  If not, see <http://www.gnu.org/licenses/>.
+// 
+#ifndef Tensor_h
+#define Tensor_h
+
+#include "typedefs.hpp"
+#include "math/Matrix.hpp"
+
+#include <unsupported/Eigen/CXX11/Tensor>
+
+
+namespace GQCP {
+
+
+/**
+ *  An extension of the Eigen::Tensor class, with extra operations
+ *
+ *  @tparam _Scalar     the scalar representation type
+ *  @tparam _Rank       the rank of the tensor, i.e. the number of indices
+ *
+ *  We have decided to inherit from Eigen::Tensor, because we will use different hierarchies: see also: https://eigen.tuxfamily.org/dox-devel/TopicCustomizing_InheritingMatrix.html
+ */
+template <typename _Scalar, int _Rank>
+class Tensor : public Eigen::Tensor<_Scalar, _Rank> {
+public:
+
+    using Scalar = _Scalar;
+    enum {
+        Rank = _Rank
+    };
+
+    using Self = Tensor<Scalar, Rank>;
+    using Base = Eigen::Tensor<Scalar, Rank>;
+
+
+public:
+
+    /*
+     *  CONSTRUCTORS
+     */
+
+    using Eigen::Tensor<Scalar, Rank>::Tensor;  // inherit base constructors
+
+
+    /*
+     *  NAMED CONSTRUCTORS
+     */
+
+    /**
+     *  @param T            a rank-4 tensor
+     *
+     *  @param i            1st starting index
+     *  @param j            2nd starting index
+     *  @param k            3rd starting index
+     *  @param l            4th starting index
+     *  @param desize       early cut-off of index iteration
+     *
+     *  @return a rank-4 tensor from an other rank-4 tensor, starting from given indices
+     */
+    template <int Z = Rank>
+    static enable_if_t<Z == 4, Self> FromBlock(const Self& T, size_t i, size_t j, size_t k, size_t l, size_t desize=0) {
+
+        Tensor<double, Rank> T_result (T.dimension(0) - i - desize, T.dimension(1) - j - desize, T.dimension(2) - k - desize, T.dimension(3) - l - desize);
+        T_result.setZero();
+
+        for (size_t p = 0; p < T_result.dimension(0); p++) {
+            for (size_t q = 0; q < T_result.dimension(1); q++) {
+                for (size_t r = 0; r < T_result.dimension(2); r++) {
+                    for (size_t s = 0; s < T_result.dimension(3); s++) {
+                        T_result(p, q, r, s) = T(i + p, j + q, k + r, l + s);
+                    }
+                }
+            }
+        }
+
+        return T_result;
+    }
+
+
+    /*
+     *  PUBLIC METHODS
+     */
+
+    /**
+     *  @return this as an Eigen::Tensor, as a work-around to fix Eigen::Tensor expressions
+     */
+    const Base& Eigen() const {
```

Comment:
Maybe we can find a better solution to this in the future

Reply:
Apparently `Eigen/Dense` did something so that we can do operations with derived classes of `Eigen::Matrix`, but that kind of support is not yet included in `Eigen/Tensor`. I hope that one day `Eigen/Tensor` does the same thing.